tvb_contrib/tvb/contrib/cosimulation/cosim_monitors.py

```diff
@@ -0,0 +1,185 @@
+# -*- coding: utf-8 -*-
+#
+#
+#  TheVirtualBrain-Contributors Package. This package holds simulator extensions.
+#  See also http://www.thevirtualbrain.org
+#
+# (c) 2012-2020, Baycrest Centre for Geriatric Care ("Baycrest") and others
+#
+# This program is free software: you can redistribute it and/or modify it under the
+# terms of the GNU General Public License as published by the Free Software Foundation,
+# either version 3 of the License, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+# PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+# You should have received a copy of the GNU General Public License along with this
+# program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+#   CITATION:
+# When using The Virtual Brain for scientific publications, please cite it as follows:
+#
+#   Paula Sanz Leon, Stuart A. Knock, M. Marmaduke Woodman, Lia Domide,
+#   Jochen Mersmann, Anthony R. McIntosh, Viktor Jirsa (2013)
+#       The Virtual Brain: a simulator of primate brain network dynamics.
+#   Frontiers in Neuroinformatics (7:10. doi: 10.3389/fninf.2013.00010)
+
+"""
+.. moduleauthor:: Lionel Kusch <lkusch@thevirtualbrain.org>
+.. moduleauthor:: Dionysios Perdikis <dionperd@gmail.com>
+"""
+
+import abc
+
+import numpy
+
+from tvb.basic.neotraits.api import HasTraits, Attr, NArray
+from tvb.simulator.coupling import Coupling, Linear
+from tvb.simulator.monitors import Raw, RawVoi, AfferentCoupling
+
+
+class CosimMonitor(HasTraits):
+    """
+    Abstract base class for cosimulation monitors implementations.
+    """
+
+    def get_sample(self, start_step, n_steps, history,cosim):
+        times = []
+        values = []
+        for step in range(start_step, start_step + n_steps):
+            if cosim:
+                state = history.query(step)
+            else:
+                state = history.query(step)[0]
+            tmp = super(self.__class__, self).sample(step, state)
+            if tmp is not None:
+                times.append(tmp[0])
+                values.append(tmp[1])
+        return [numpy.array(times), numpy.array(values)]
+
+    @abc.abstractmethod
+    def sample(self, start_step, n_steps, cosim_history, history):
+        """
+        This method provides monitor output, and should be overridden by subclasses.
+        Use the original signature.
+        """
+        pass
+
+class CosimMonitorFromCoupling(CosimMonitor):
+    """
+       Abstract base class for a monitor that records the future coupling values.
+       !!!WARNING don't use this monitor for a time smaller than the synchronization variable!!!
```

Comment:
Docstring warnings will not show up in a debugger; please insert appropriate `assert`s in the initialization or `get_sample` method to ensure the use is correct.

Reply:
I have removed the warning from the docstrings.
The ideal place for a check is upon configuration, since then simulator.synchronization_time will be available, e.g., by overwriting the Monitor._config_time() method. 
However, I am not sure what kind of assertion is due here.
@lionelkusch could you help with this?

Reply:
I did the modification about this comment in this commit : 28dac89
I am not sure it's the best way. If you have more comment about do it.

Reply:
That commit adds a current_step argument to a bunch of methods, but there's no assert or exception raised, which is what I meant, something like 
```python
if t0 < t1:
    raise ValueError("don't use this monitor for a time smaller than the synchronization variable")
```

Reply:
I add all the exception in the abstract class.